### ISSUE TITLE
memref.alloca / copy ops are created for constants in IREEComprehensiveBufferize

### ISSUE THREAD

**Issue by hanhanW** [google/iree#8660]
The generic op takes constants in outs. The constant op is lowered to memref.get_global, and an memref.alloca op is created. I thought we don't need memref.alloca ops in this case, but I was wrong. The constant is read-only, so we need an extra buffer.

I triaged and found that an upstream canonicalization pattern introduces such pattern. `linalg.fill + linalg.generic` is folded into `linalg.generic` op. https://reviews.llvm.org/D121535

IIUC, we should not fold fill op into generic op in this case. If we don't fold them into one generic op, we'll have proper bufferization. Because the outs would be output buffer, not constants. @matthias-springer @MaheshRavishankar  please verify if I'm correct about bufferization issue.

@MaheshRavishankar @ThomasRaoux @nirvedhmeshram I think the canonicalization pattern is problematic, do we want to revert it? How do we fix the issue?

IR before bufferization:

```mlir
func @cumsum_f32_dispatch_0() {
  %c2 = arith.constant 2 : index
  %c0 = arith.constant 0 : index
  %cst = arith.constant dense<0.000000e+00> : tensor<1x1x2xf32>
  %0 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) offset(%c0) alignment(64) : !flow.dispatch.tensor<readonly:3x2x2xf32>
  %1 = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) offset(%c0) alignment(64) : !flow.dispatch.tensor<writeonly:2x2x2xf32>
  %workgroup_id_x = hal.interface.workgroup.id[0] : index
  %workgroup_count_x = hal.interface.workgroup.count[0] : index
  %workgroup_id_y = hal.interface.workgroup.id[1] : index
  %workgroup_count_y = hal.interface.workgroup.count[1] : index
  %workgroup_id_z = hal.interface.workgroup.id[2] : index
  %workgroup_count_z = hal.interface.workgroup.count[2] : index
  %2 = affine.apply affine_map<()[s0] -> (s0 * 2)>()[%workgroup_id_x]
  %3 = affine.apply affine_map<()[s0] -> (s0 * 2)>()[%workgroup_count_x]
  %4 = linalg.init_tensor [2] : tensor<2xf32>
  scf.for %arg0 = %workgroup_id_z to %c2 step %workgroup_count_z {
    %5 = affine.min affine_map<(d0) -> (2, -d0 + 3)>(%arg0)
    scf.for %arg1 = %workgroup_id_y to %c2 step %workgroup_count_y {
      scf.for %arg2 = %2 to %c2 step %3 {
        %6 = flow.dispatch.tensor.load %0, offsets = [%arg0, %arg1, %arg2], sizes = [%5, 1, 2], strides = [1, 1, 1] : !flow.dispatch.tensor<readonly:3x2x2xf32> -> tensor<?x1x2xf32>
        %7 = tensor.extract_slice %6[0, 0, 0] [2, 1, 2] [1, 1, 1] : tensor<?x1x2xf32> to tensor<2x1x2xf32>
        %8 = linalg.generic {indexing_maps = [affine_map<(d0, d1, d2, d3) -> (d0 + d3, d1, d2)>, affine_map<(d0, d1, d2, d3) -> (d3)>, affine_map<(d0, d1, d2, d3) -> (d0, d1, d2)>], iterator_types = ["parallel", "parallel", "parallel", "reduction"]} ins(%7, %4 : tensor<2x1x2xf32>, tensor<2xf32>) outs(%cst : tensor<1x1x2xf32>) attrs =  {lowering_config = #iree_codegen.lowering_config<tile_sizes = [[1, 1, 2, 0], [1, 1, 2, 0], [0, 0, 0, 16]]>} {
        ^bb0(%arg3: f32, %arg4: f32, %arg5: f32):
          %9 = arith.addf %arg3, %arg5 : f32
          linalg.yield %9 : f32
        } -> tensor<1x1x2xf32>
        flow.dispatch.tensor.store %8, %1, offsets = [%arg0, %arg1, %arg2], sizes = [1, 1, 2], strides = [1, 1, 1] : tensor<1x1x2xf32> -> !flow.dispatch.tensor<writeonly:2x2x2xf32>
      }
    }
  }
  return
}
```

IR after bufferization:

```mlir
func @cumsum_f32_dispatch_0() {
  %c0 = arith.constant 0 : index
  %c2 = arith.constant 2 : index
  %0 = memref.get_global @__constant_1x1x2xf32 : memref<1x1x2xf32>
  %1 = memref.alloca() {alignment = 128 : i64} : memref<1x1x2xf32>
  %2 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) offset(%c0) alignment(64) : memref<3x2x2xf32>
  memref.assume_alignment %2, 64 : memref<3x2x2xf32>
  %3 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) offset(%c0) alignment(64) : !flow.dispatch.tensor<readonly:3x2x2xf32>
  %4 = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) offset(%c0) alignment(64) : memref<2x2x2xf32>
  memref.assume_alignment %4, 64 : memref<2x2x2xf32>
  %5 = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) offset(%c0) alignment(64) : !flow.dispatch.tensor<writeonly:2x2x2xf32>
  %workgroup_id_x = hal.interface.workgroup.id[0] : index
  %workgroup_count_x = hal.interface.workgroup.count[0] : index
  %workgroup_id_y = hal.interface.workgroup.id[1] : index
  %workgroup_count_y = hal.interface.workgroup.count[1] : index
  %workgroup_id_z = hal.interface.workgroup.id[2] : index
  %workgroup_count_z = hal.interface.workgroup.count[2] : index
  %6 = affine.apply affine_map<()[s0] -> (s0 * 2)>()[%workgroup_id_x]
  %7 = affine.apply affine_map<()[s0] -> (s0 * 2)>()[%workgroup_count_x]
  %8 = memref.alloca() {alignment = 128 : i64} : memref<2xf32>
  scf.for %arg0 = %workgroup_id_z to %c2 step %workgroup_count_z {
    %9 = affine.min affine_map<(d0) -> (2, -d0 + 3)>(%arg0)
    scf.for %arg1 = %workgroup_id_y to %c2 step %workgroup_count_y {
      scf.for %arg2 = %6 to %c2 step %7 {
        %10 = memref.subview %2[%arg0, %arg1, %arg2] [%9, 1, 2] [1, 1, 1] : memref<3x2x2xf32> to memref<?x1x2xf32, affine_map<(d0, d1, d2)[s0] -> (d0 * 4 + s0 + d1 * 2 + d2)>>
        %11 = memref.subview %10[0, 0, 0] [2, 1, 2] [1, 1, 1] : memref<?x1x2xf32, affine_map<(d0, d1, d2)[s0] -> (d0 * 4 + s0 + d1 * 2 + d2)>> to memref<2x1x2xf32, affine_map<(d0, d1, d2)[s0] -> (d0 * 4 + s0 + d1 * 2 + d2)>>
        linalg.generic {indexing_maps = [affine_map<(d0, d1, d2) -> (d0, d1, d2)>, affine_map<(d0, d1, d2) -> (d0, d1, d2)>], iterator_types = ["parallel", "parallel", "parallel"]} ins(%0 : memref<1x1x2xf32>) outs(%1 : memref<1x1x2xf32>) {
        ^bb0(%arg3: f32, %arg4: f32):
          linalg.yield %arg3 : f32
        }
        linalg.generic {indexing_maps = [affine_map<(d0, d1, d2, d3) -> (d0 + d3, d1, d2)>, affine_map<(d0, d1, d2, d3) -> (d3)>, affine_map<(d0, d1, d2, d3) -> (d0, d1, d2)>], iterator_types = ["parallel", "parallel", "parallel", "reduction"]} ins(%11, %8 : memref<2x1x2xf32, affine_map<(d0, d1, d2)[s0] -> (d0 * 4 + s0 + d1 * 2 + d2)>>, memref<2xf32>) outs(%1 : memref<1x1x2xf32>) attrs =  {lowering_config = #iree_codegen.lowering_config<tile_sizes = [[1, 1, 2, 0], [1, 1, 2, 0], [0, 0, 0, 16]]>} {
        ^bb0(%arg3: f32, %arg4: f32, %arg5: f32):
          %13 = arith.addf %arg3, %arg5 : f32
          linalg.yield %13 : f32
        }
        %12 = memref.subview %4[%arg0, %arg1, %arg2] [1, 1, 2] [1, 1, 1] : memref<2x2x2xf32> to memref<1x1x2xf32, affine_map<(d0, d1, d2)[s0] -> (d0 * 4 + s0 + d1 * 2 + d2)>>
        linalg.generic {indexing_maps = [affine_map<(d0, d1, d2) -> (d0, d1, d2)>, affine_map<(d0, d1, d2) -> (d0, d1, d2)>], iterator_types = ["parallel", "parallel", "parallel"]} ins(%1 : memref<1x1x2xf32>) outs(%12 : memref<1x1x2xf32, affine_map<(d0, d1, d2)[s0] -> (d0 * 4 + s0 + d1 * 2 + d2)>>) {
        ^bb0(%arg3: f32, %arg4: f32):
          linalg.yield %arg3 : f32
        }
      }
    }
  }
  return
}

```

**Comment by MaheshRavishankar** [comment]
The `fill` -> `generic` shouldnt have been a canonicalization pattern, but leaving that aside bufferization should still be accounting for this case. Issue seems to be that the `linalg.init_tensor` is not removed by the pass to convert to destination passing style. Can you post the IR before and after that?

**Comment by hanhanW** [comment]
Nope, the init_tensor is shaped only operand. That would be killed after lowering to loops. The issue here is:

1. We allocate a stack buffer.
2. Copy constants to stack buffer.
3. Use the stack buffer in outs
4. Copy the result from stack buffer to output buffer.

My point is `fill + generic` -> `generic` causes this issue. IMO, this is a wrong pattern because constants have read-only semantic, but init_tensor have read-write semantic. We can not simply replace it with constant.

**Comment by hanhanW** [comment]
Note that this is a reduce_window op.

Here is the full IR dump: https://gist.githubusercontent.com/hanhanW/1bd107dbafb51c4be39683e6ecb1703a/raw

**Comment by MaheshRavishankar** [comment]
> Nope, the init_tensor is shaped only operand. That would be killed after lowering to loops. The issue here is:
Ah, didnt realize that. Ok cool.

> 
> 1. We allocate a stack buffer.
> 2. Copy constants to stack buffer.
> 3. Use the stack buffer in outs
> 4. Copy the result from stack buffer to output buffer.
> 
> My point is `fill + generic` -> `generic` causes this issue. IMO, this is a wrong pattern because constants have read-only semantic, but init_tensor have read-write semantic. We can not simply replace it with constant.

In theory, `init_tensor` has no read/write/read-write semantics. It is just carrying the shape of the result. So the `generic` op could use `%cst` since it is only looking at the shape.
The way upstream bufferization uses `init_tensor` and interprets constant is sort of the issue here (though it is still respecting the "contract"). So there are two solutions I see

1) Move the pattern you pointed into `ElementwiseOpFusion` so it triggers only at that stage. IMO that is the right place for it. Though this is a WAR
2) The convert to destination passing style needs a separate "fix up heuristic" to make `linalg.generic` that takes a "read-only" tensor as the `outs` operand (instead of an `init_tensor`). To use an `init_tensor` instead. This is really a "fix up". So the convert to destination passing style could do as its "first step" (before even analysis), a pass to replace all `linalg.generic` ops where the `outs` isnt used with `linalg.init_tensor`. That creates a clean slate for everything else to work as expected.

**Comment by nirvedhmeshram** [comment]
@hanhanW If you choose option 1, I can do it. Let me know. @MaheshRavishankar for option 1, would you also move the companion pattern that removes dead input arguments as long as there is no dimension information lost? In my opinion that's a reasonable canonicalization.

**Comment by hanhanW** [comment]
I see, I misunderstood init_tensor semantic. Thanks for the explanation!

I'd prefer not generating this pattern in the first place, so we don't need a fix up. Moving the pattern into elementwise op fusion makes sense to me. @nirvedhmeshram please help on this, thank you!

**Comment by MaheshRavishankar** [comment]
@hanhanW and @nirvedhmeshram I'd implement both. 
1) Moving the `linalg.fill` -> `linalg.generic` pattern into fusion
2) Break the false dependence between `arith.constant` -> `linalg.generic`. The pattern already exists ([here](https://github.com/llvm/llvm-project/blob/45e85feba6e49a4d3257b78f1a50f94b3c79a492/mlir/lib/Dialect/Linalg/Transforms/ElementwiseOpFusion.cpp#L2173)). It isnt added as a canonicalization since there were concerns about how it interacts with bufferization. But it could be used as a "pre-processing" in convert to destination passing style pass.

Also @nirvedhmeshram you can keep the pattern to drop unused arguments as a canonicalizations.

**Comment by hanhanW** [comment]
Okay, thanks for the pointer. I'll study it!

**Comment by nirvedhmeshram** [comment]
for the fill->genric pattern Is it okay to put the pattern in its own anonymous namespace? If I try to add it to anonymous namespaces of existing patterns it gives some errors.

**Comment by MaheshRavishankar** [comment]
That shouldnt happen. Can you give some more details about the error?

**Comment by nirvedhmeshram** [comment]
Nevermind, I had a mistake with the braces,
@MaheshRavishankar could you please take a look at this commit, if it looks okay, I can submit a diff and add you as a reviewer
https://github.com/nirvedhmeshram/llvm-project/commit/4db543b4db40b4d8940aca4f598352b49e033f78

**Comment by MaheshRavishankar** [comment]
Yeah, go ahead and just send the diff for review on phabricator...

**Comment by hanhanW** [comment]
Thanks @nirvedhmeshram, the fix has landed into IREE.

I checked the IR again and found it's still the case. I found that it's "wrong" when it's lowered from mhlo ops. Different from other reduce ops, it broadcasts the init_value (which is a scalar constant) and use it in `outs`. This introduces the pattern at the very beginning stage.

https://github.com/tensorflow/tensorflow/blob/6fdf81d0592bc4f91ad39d1d62383e06671a6f81/tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc#L2207-L2216

I verified that it works if we can replace broadcast op with `init_tensor + fill` op with your patch. I'll take care the rest, thanks for helping on this!

RE `RemoveOutsDependency` pattern:
The pattern only works if `outs` operand is not used within the op. In this case, it's used. I think I can add a similar pattern. If there are splat constants in outs, we can replace it with `init_tensor + fill`.
